yt_dlp/extractor/youtube.py

```diff
@@ -4031,16 +4098,39 @@ def append_client(*client_names):
 
             visitor_data = visitor_data or self._extract_visitor_data(master_ytcfg, initial_pr, player_ytcfg)
             data_sync_id = data_sync_id or self._extract_data_sync_id(master_ytcfg, initial_pr, player_ytcfg)
-            po_token = self.fetch_po_token(
-                client=client, visitor_data=visitor_data,
-                data_sync_id=data_sync_id if self.is_authenticated else None,
-                player_url=player_url if require_js_player else None,
-            )
 
-            require_po_token = self._get_default_ytcfg(client).get('REQUIRE_PO_TOKEN')
-            if not po_token and require_po_token and 'missing_pot' in self._configuration_arg('formats'):
+            fetch_po_token_args = {
+                'client': client, 'visitor_data': visitor_data, 'video_id': video_id,
+                'data_sync_id': data_sync_id if self.is_authenticated else None,
+                'player_url': player_url if require_js_player else None,
+                'session_index': self._extract_session_index(master_ytcfg, player_ytcfg),
+                'ytcfg': player_ytcfg,
+            }
+
+            player_po_token = self.fetch_po_token(
+                context=_PoTokenContext.PLAYER, **fetch_po_token_args)
+
+            gvs_po_token_func = lambda: self.fetch_po_token(
+                context=_PoTokenContext.GVS, **fetch_po_token_args)
+
+            # Fetch an initial PO Token for GVS upfront
+            gvs_po_token = gvs_po_token_func()
+
+            require_player_po_token = _PoTokenContext.PLAYER in self._get_default_ytcfg(client).get('PO_TOKEN_REQUIRED_CONTEXTS')
+            if not player_po_token and require_player_po_token:
```

Comment:
since we `setdefault` for all clients, we don't need `.get()` and IMO we'd want a key error here to notice if we change things elsewhere and forget to here. I think it's also less taxing to read if we use a var for the required contexts instead of temp var for `require_player_po_token` since we do the same thing below
```suggestion
            po_token_required_contexts = self._get_default_ytcfg(client)['PO_TOKEN_REQUIRED_CONTEXTS']

            if not player_po_token and _PoTokenContext.PLAYER in po_token_required_contexts:
```